include/deal.II/fe/fe_series.h

```diff
@@ -0,0 +1,428 @@
+     */
+    const hp::QCollection<dim>  &q_collection;
+
+    /**
+     * Ensure that the transformation matrix for FiniteElement index
+     * @p fe_index is calculated. If not, calculate it.
+     */
+    void ensure_existence(const unsigned int fe_index);
+
+    /**
+     * Transformation matrices for each FiniteElement.
+     */
+    std::vector<FullMatrix<double> > legendre_transform_matrices;
+
+    /**
+     * Auxiliary vector to store unrolled coefficients.
+     */
+    std::vector<double> unrolled_coefficients;
+
+  };
+
+
+  /**
+   * Calculate the @p norm of subsets of @p coefficients defined by
+   * @p predicate being constant. Return the pair of vectors of predicate values
+   * and the vector of calculated subset norms.
+   *
+   * @p predicate should return a pair of <code>bool</code> and <code>unsigned int</code>.
+   * The former is a flag whether a given TableIndices should be used in
+   * calculation, whereas the latter is the unrolled value of indices according
+   * to which the subsets of coefficients will be formed.
+   *
+   * @note Only the following values of @p norm are implemented and make sense
+   * in this case: mean, L1_norm, L2_norm, Linfty_norm. The mean norm can only
+   * be applied to real valued coefficients.
+   */
+  template <int dim, typename T>
+  std::pair<std::vector<unsigned int>,std::vector<double> >
+  process_coefficients(const Table<dim,T> &coefficients,
+                       const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<dim> &)> &predicate,
+                       const VectorTools::NormType norm);
+
+
+
+  /**
+   * Linear regression least-square fit of $y = k \, x + b$.
+   * The size of the input vectors should be equal and more than 1.
+   * The returned pair will contain $k$ (first) and $b$ (second).
+   */
+  std::pair<double,double> linear_regression(const std::vector<double> &x,
+                                             const std::vector<double> &y);
+
+}
+
+/*@}*/
+
+#ifndef DOXYGEN
+
+// -------------------  inline and template functions ----------------
+
+namespace
+{
+  template <int dim,typename T>
+  void fill_map_index(const Table<dim,T> &coefficients,
+                      const TableIndices<dim> &ind,
+                      const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<dim> &)> &predicate,
+                      std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    const std::pair<bool,unsigned int> pred_pair = predicate(ind);
+    // don't add a value if predicate is false
+    if (pred_pair.first == false)
+      return;
+
+    const unsigned int &pred_value = pred_pair.second;
+    const T &coeff_value = coefficients(ind);
+
+    if (pred_to_values.find (pred_value) == pred_to_values.end())
+      {
+        pred_to_values[pred_value] = std::vector<T>(1,coeff_value);
+      }
+    else
+      {
+        pred_to_values[pred_value].push_back(coeff_value);
```

Comment:
if an entry does not exist, `pred_to_values[pred_value]` will create it, otherwise it returns the existing entry. As far as I can tell, the code above should be ok.